src/logd.rs

```diff
@@ -0,0 +1,117 @@
+use std::{io, os::unix::net::UnixDatagram, path::Path, time::SystemTime};
+
+use bytes::BufMut;
+use parking_lot::RwLockUpgradableReadGuard;
+
+use crate::{thread, Buffer, Event, Priority, LOGGER_ENTRY_MAX_LEN};
+
+/// Logd write socket path
+const LOGDW: &str = "/dev/socket/logdw";
+
+lazy_static::lazy_static! {
+    static ref SOCKET: LogdSocket = LogdSocket::connect(Path::new(LOGDW));
+}
+
+/// Logd write socket abstraction. Sends never fail and on each send a reconnect
+/// reconnect attempt is made.
+struct LogdSocket {
+    socket: parking_lot::RwLock<UnixDatagram>,
+}
+
+impl LogdSocket {
+    /// Construct a new LogdSocket.
+    ///
+    /// # Panics
+    ///
+    /// This function panics when the socket cannot be created.
+    pub fn connect(path: &Path) -> LogdSocket {
+        let socket = UnixDatagram::unbound().expect("Failed to create socket");
+        socket.connect(path).ok();
+        let lock = parking_lot::RwLock::new(socket);
+        LogdSocket { socket: lock }
+    }
+
+    /// Write a log entry to the log daemon. If a first write attempt fails, try to
+    /// reconnect to the log daemon and try again.
+    pub fn send(&self, buffer: &[u8]) -> io::Result<()> {
+        let lock = self.socket.upgradable_read();
+        if lock.send(buffer).is_err() {
+            // Try to crate an unbounded socket. Expect this to work.
+            let socket = UnixDatagram::unbound()?;
+
+            // Upgrade the read lock and replace the socket if the sent attempt is successful.
+            let mut lock = RwLockUpgradableReadGuard::upgrade(lock);
+            socket.connect(LOGDW)?;
+            socket.send(buffer)?;
+            // Assign the new socket to the lock. In the worst case one or more threads
+            // are opening sockets to logd which are immediately closed.
+            *lock = socket;
+        }
```

Comment:
I like the mutex trick to handle the new connection. Just curious about how frequent does this happen and if a single reconnect is enough or if a classic retry loop should be used?